### PR TITLE
Support multi-tenancy for process instance creation in the engine

### LINKED ISSUE #13279: Support multi-tenancy for core BPMN process operations
### Description

This issue covers multi-tenancy support for the core operations of BPMN processes in Zeebe. The following should be covered, i.e. any command/event records produced by the following actions should contain a `tenantId`:
* Starting a process instance
* Starting a process instance with results
* Starting a CallActivitiy
* Evaluating a Decision through a Business Rule task.
* Child instance creation.

#### Expected behaviors
The following behavior should be supported:
* A process instance is always owned by a tenant. A process instance can’t be shared across tenants.
* All child instances are owned by the same tenant as the process instance. That means that any `ProcessInstanceRelated` records are owned by the same tenant as the process instance.
* A `CallActivity` is started from a process definition with the following criteria:
   1. The process definition is owned by the same tenant as the parent process instance.
   2. If nothing is found, an `Incident` is raised.
* A `Decision` is evaluated if it has the following criteria:
   1. The Decision is owned by the same tenant as the parent process instance.
   2. If nothing is found, an `Incident` is raised.
* Filtering process definitions by version remains the same as the non-multi-tenant approach, i.e. if no version is specified, the latest version of the process definition owned by the tenant is used. (provided by #13238)
* A process instance can be started only if a tenant id is provided with the client request. The following scenarios are possible:
   * A tenant id is provided:
      1. A process instance is started if the user making the request has access to the tenant owning the process definition.
      2. The request/command is rejected if the user doesn't have correct access to the tenant.
   * A tenant id isn't provided:
      1. A process instance is started if the user only has access to a single tenant, and a single process definition can be found that can be mapped to that tenant.
      2. The request/command is rejected if the user has access to multiple tenants or no process definition is found.

:information_source: In the implementation, first focus on implementing the simple path, with a `tenantId` explicitly provided in the client request/command record. The goal is to deliver an MVP as soon as possible.

```[tasklist]
### Task breakdown
- [x] Engine can process multi-tenant ProcessInstanceCreationRecord commands with tenant id
- [x] Engine propagates tenant id to process instance elements.
- [x] Engine can start CallActivity by tenant id
- [ ] https://github.com/camunda/zeebe/issues/14211
- [ ] https://github.com/camunda/zeebe/issues/13536
- [ ] Notify Operate and Optimize teams that core BPMN process events are available from Zeebe
- [ ] ~CreateProcessInstance* RPCs documentation is expanded with multi-tenancy~
- [ ] ~Engine can evaluate Business Rule task Decision by tenant id~ (moved to DMN)
- [ ] ~Multi-tenancy data migration is provided for `ElementInstanceState`, `BannedInstanceState`~ WON'T DO
```

### Blocked by
- https://github.com/camunda/zeebe/issues/13238

### PR DESCRIPTION
**Pull request by koevskinikola**
## Description

Support multi-tenancy for process instance creation in the engine.

## Related issues

closes #13279 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by koevskinikola** [comment]
@remcowesterhoud commits f96602f6f7c2ce3db60c65a899e94a7daa526f0e and 8dc635e49842a42302d2ece0e06c5bb327c97bd5 forward the `tenantId` to the `VariableBehavior` class so that it's ready to be used when we start #13388.

**Comment by remcowesterhoud** [comment]
@koevskinikola Fyi, there's some failing test cases. I'll review it anyway but wanted to notify you early.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java [hunk 1]
```diff
@@ -287,8 +287,7 @@ private void unsubscribeFromPreviousTimer(
       final ProcessMetadata processMetadata, final TimerInstance timer) {
     final DirectBuffer timerBpmnId =
         processState
-            .getProcessByKeyAndTenant(
-                timer.getProcessDefinitionKey(), processMetadata.getTenantId())
+            .getProcessByKeyAndTenant(timer.getProcessDefinitionKey(), timer.getTenantId())
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java:290]: 👍 Good catch

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/TimerInstance.java [hunk 2]
```diff
@@ -25,6 +28,8 @@ public final class TimerInstance extends UnpackedObject implements DbValue {
   private final LongProperty keyProp = new LongProperty("key", 0L);
   private final LongProperty elementInstanceKeyProp = new LongProperty("elementInstanceKey", 0L);
   private final LongProperty processInstanceKeyProp = new LongProperty("processInstanceKey", 0L);
+  private final StringProperty tenantIdProp =
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/instance/TimerInstance.java:31]: ❌ You forgot to declare the property in the constructor

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java [hunk 3]
```diff
@@ -49,11 +49,11 @@ public final class CreateProcessInstanceProcessor
   private static final String ERROR_MESSAGE_NO_IDENTIFIER_SPECIFIED =
       "Expected at least a bpmnProcessId or a key greater than -1, but none given";
   private static final String ERROR_MESSAGE_NOT_FOUND_BY_PROCESS =
-      "Expected to find process definition with process ID '%s', but none found";
+      "Expected to find process definition with process ID '%s' and tenant ID '%s', but none found";
   private static final String ERROR_MESSAGE_NOT_FOUND_BY_PROCESS_AND_VERSION =
-      "Expected to find process definition with process ID '%s' and version '%d', but none found";
+      "Expected to find process definition with process ID '%s',  tenant ID '%s', and version '%d', but none found";
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java:54]: 🙃 you put two whitespaces in the message

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Nice work @koevskinikola! You've missed to declare 1 property and there's some failing tests. Otherwise it LGTM!

**Comment by koevskinikola** [comment]
@remcowesterhoud I implemented your feedback, and the CI is passing now.

**Review by remcowesterhoud — APPROVED** [review]
Awesome, thanks! 🚀 

bors merge

**Pull request merged**